### ISSUE TITLE
Redfish support for Host HA (Out of Band)

### ISSUE THREAD

**Issue by wido** [apache/cloudstack#3624]
##### ISSUE TYPE
* Feature Idea

##### COMPONENT NAME
~~~
Management Server
~~~

##### CLOUDSTACK VERSION
~~~
master
~~~

##### SUMMARY
The current implementation of Host HA (Out of Band) only supports IPMI (v2).

- https://cwiki.apache.org/confluence/display/CLOUDSTACK/Host+HA
- https://www.shapeblue.com/host-ha-for-kvm-hosts-in-cloudstack/

Redfish is a new implementation supported by the big vendors which provides a HTTP REST API to control servers.

<pre>
DMTF’s Redfish® is a standard designed to deliver simple and secure management for converged, hybrid IT and the Software Defined Data Center (SDDC). Both human readable and machine capable, Redfish leverages common Internet and web services standards to expose information directly to the modern tool chain.
</pre>

- https://www.dmtf.org/standards/redfish
- https://en.wikipedia.org/wiki/Redfish_(specification)

The follow manufacturers support Redfish in their OOB solutions:

- Dell's iDRAC
- HPE's ILO
- SuperMicro's IPMI (>X10 generation)

By supporting Redfish we can support the latest generation of hardware and specifications for the OOB management of Hosts.

There is no need to use 'ipmitool' as Redfish are just HTTP calls which need to be send to the server's OOB system to reboot the Host when needed.

**Comment by wido** [comment]
![Screenshot from 2019-10-07 10-15-59](https://user-images.githubusercontent.com/326786/66295493-c1ecf000-e8eb-11e9-8c02-e80edc6554cd.png)

There should be a Redfish driver there

**Comment by wido** [comment]
SuperMicro: https://www.supermicro.com/en/solutions/management-software/redfish
Dell: https://www.dell.com/support/article/in/en/inbsd1/sln310624/redfish

**Comment by andrijapanicsb** [comment]
Moving this one to 4.15, I don't see it happening in 4.14 due to lack of time.

**Comment by wido** [comment]
Work is in progress here. @GabrielBrascher has been writing a lot of code: https://github.com/PCextreme/cloudstack/commits/redfish-oob-driver

**Comment by GabrielBrascher** [comment]
**[Update]** On the last couple days we did some progress in this implementation.

Tested Redfish Client and Redfish OOB management driver on nodes with Dell iDRAC8 (thanks to our folks at [Yourhosting](www.yourhosting.nl)). We still going to run a few tests on Supermicro nodes as well as there are a few differences on the API calls from Dell to Supermicro Rest API.

Here follow a few pics.
![redfish1](https://user-images.githubusercontent.com/5025148/79600831-2e785100-80be-11ea-9ef5-fe5885f56373.jpg)
![redfish2](https://user-images.githubusercontent.com/5025148/79600841-2fa97e00-80be-11ea-97fd-e5b07e3ff157.jpg)

We have been able to test and it works fine:

1. via the CloudStack UI, enable redfish oob driver for host n05
2. verified that ACS can retrieve the redfish system ID and system power state
3. powering Off node works as expected, host state transition worked well and the host was indeed powered off when verifying via the idrac UI
4. powering On also worked as expected, CloudStack host state transition as expected, host is indeed up again.

It is still necessary to do a few enhancements on the code to have it ready for a PR (e.g. Java docs, and adding unit tests). As soon as we are ready to open a PR I will link the PR here.
